File: contests/abc089/d.cpp
#include<iostream>
#include<vector>
#include<map>
#include<utility>
using namespace std;

vector<pair<int, int> > table(300*300+1);
map<string, int> cache;

int calc(int l, int r, int d) {
  if (l==r) {
    return 0;
  }

  string k = to_string(l)+"|"+to_string(r);
  if (cache.find(k) != cache.end()) {
    return cache[k];
  }
  int ret = abs(table[l].first - table[l+d].first) + abs(table[l].second - table[l+d].second) + calc(l+d,r,d);
  cache[k] = ret;
  return ret;
}

int main() {
  int h,w,d;
  cin>>h>>w>>d;

  for (int hi=0;hi<h;++hi) {
    for (int wi=0;wi<w;++wi) {
      int x;
      cin>>x;
      table[x] = make_pair(hi+1, wi+1);
    }
  }

  int q;
  cin>>q;
  for (int i=0;i<q;++i) {
    int l,r;
    cin>>l>>r;
    cout<<calc(l,r,d)<<endl;
  }
}

File: contests/abc089/d.py
# coding=utf-8
line = input().split(" ")
h = int(line[0])
w = int(line[1])
d = int(line[2])

table = [0 for _ in range(h*w+1)]
for hi in range(h):
    line = list(map(int, input().split(" ")))

    for wi in range(w):
        table[line[wi]] = (hi+1, wi+1)

def calc(l,r):
    ret = 0
    if l!=r:
        c1 = table[l]
        c2 = table[r]
        ret += abs(c1[0]-c2[0])+abs(c1[1]-c2[1])
    return ret

acc = [0 for _ in range(h*w+1)]
for dd in range(1,d+1):
    s = dd+d
    while s<=h*w:
        acc[s] = calc(s-d,s)+acc[s-d]
        s += d

q = int(input())
for _ in range(q):
    line = input().split(" ")
    l = int(line[0])
    r = int(line[1])

    ret = acc[r] - acc[l]
    print(ret)
